File: backend/app/services/ai_provider_service.py
"""
AI Provider Service
Abstraction layer for multiple AI providers (OpenAI, Groq, Gemini)
"""

import os
import logging
from typing import List, Dict, Any, Optional
from openai import AsyncOpenAI
from groq import AsyncGroq
import google.generativeai as genai

logger = logging.getLogger(__name__)


class AIProviderService:
    """Unified interface for multiple AI providers"""

    def __init__(self):
        # Initialize API clients with keys from environment (Doppler)
        self.openai_client = None
        self.groq_client = None
        self.gemini_model = None

        # Load OpenAI
        if os.getenv("OPENAI_API_KEY"):
            try:
                self.openai_client = AsyncOpenAI(api_key=os.getenv("OPENAI_API_KEY"))
                logger.info("✅ OpenAI provider initialized")
            except Exception as e:
                logger.error(f"Failed to initialize OpenAI: {e}")

        # Load Groq
        if os.getenv("GROQ_API_KEY"):
            try:
                self.groq_client = AsyncGroq(api_key=os.getenv("GROQ_API_KEY"))
                logger.info("✅ Groq provider initialized")
            except Exception as e:
                logger.error(f"Failed to initialize Groq: {e}")

        # Load Gemini
        if os.getenv("GEMINI_API_KEY"):
            try:
                genai.configure(api_key=os.getenv("GEMINI_API_KEY"))
                self.gemini_model = genai.GenerativeModel('gemini-1.5-pro')
                logger.info("✅ Gemini provider initialized")
            except Exception as e:
                logger.error(f"Failed to initialize Gemini: {e}")

    def is_provider_available(self, provider: str) -> bool:
        """Check if a provider is available"""
        if provider == "openai":
            return self.openai_client is not None
        elif provider == "groq":
            return self.groq_client is not None
        elif provider == "gemini":
            return self.gemini_model is not None
        return False

    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        tools: Optional[List[Dict[str, Any]]] = None,
        preferred_model: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: Optional[int] = None,
        stream: bool = False,
        user_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Unified chat completion with automatic fallback

        Returns:
            {
                "content": str,
                "model": str,
                "provider": str,
                "tool_calls": Optional[List],
                "usage": {"prompt_tokens": int, "completion_tokens": int, "total_tokens": int},
                "finish_reason": str,
                "cost_usd": float
            }
        """
        # Determine provider order
        providers_to_try = self._get_provider_order(preferred_model)

        last_error = None

        for provider in providers_to_try:
            try:
                if provider == "openai" and self.openai_client:
                    return await self._call_openai(messages, tools, temperature, max_tokens)
                elif provider == "groq" and self.groq_client:
                    return await self._call_groq(messages, tools, temperature, max_tokens)
                elif provider == "gemini" and self.gemini_model:
                    return await self._call_gemini(messages, temperature, max_tokens)
            except Exception as e:
                logger.warning(f"{provider} failed: {e}, trying next provider...")
                last_error = e
                continue

        # All providers failed
        raise Exception(f"All AI providers failed. Last error: {last_error}")

    def _get_provider_order(self, preferred: Optional[str]) -> List[str]:
        """Get ordered list of providers to try"""
        if preferred:
            # Try preferred first, then others
            order = [preferred]
            for p in ["openai", "groq", "gemini"]:
                if p != preferred:
                    order.append(p)
            return order

        # Default order: OpenAI (reliable) → Groq (free) → Gemini (fallback)
        return ["openai", "groq", "gemini"]

    async def _call_openai(
        self,
        messages: List[Dict[str, str]],
        tools: Optional[List[Dict[str, Any]]],
        temperature: float,
        max_tokens: Optional[int]
    ) -> Dict[str, Any]:
        """Call OpenAI API"""
        params = {
            "model": "gpt-4o-mini",
            "messages": messages,
            "temperature": temperature,
        }

        if max_tokens:
            params["max_tokens"] = max_tokens

        if tools:
            # Convert to OpenAI tool format
            params["tools"] = [
                {
                    "type": "function",
                    "function": {
                        "name": tool["name"],
                        "description": tool["description"],
                        "parameters": tool["parameters"]
                    }
                } for tool in tools
            ]

        response = await self.openai_client.chat.completions.create(**params)

        # Extract tool calls if any
        tool_calls = None
        if response.choices[0].message.tool_calls:
            tool_calls = [
                {
                    "id": tc.id,
                    "name": tc.function.name,
                    "arguments": tc.function.arguments
                } for tc in response.choices[0].message.tool_calls
            ]

        # Calculate cost (GPT-4o-mini pricing)
        prompt_tokens = response.usage.prompt_tokens
        completion_tokens = response.usage.completion_tokens
        cost_usd = (prompt_tokens * 0.150 / 1_000_000) + (completion_tokens * 0.600 / 1_000_000)

        return {
            "content": response.choices[0].message.content or "",
            "model": "gpt-4o-mini",
            "provider": "openai",
            "tool_calls": tool_calls,
            "usage": {
                "prompt_tokens": prompt_tokens,
                "completion_tokens": completion_tokens,
                "total_tokens": response.usage.total_tokens
            },
            "finish_reason": response.choices[0].finish_reason,
            "cost_usd": cost_usd
        }

    async def _call_groq(
        self,
        messages: List[Dict[str, str]],
        tools: Optional[List[Dict[str, Any]]],
        temperature: float,
        max_tokens: Optional[int]
    ) -> Dict[str, Any]:
        """Call Groq API"""
        params = {
            "model": "llama-3.3-70b-versatile",
            "messages": messages,
            "temperature": temperature,
        }

        if max_tokens:
            params["max_tokens"] = max_tokens

        if tools:
            params["tools"] = [
                {
                    "type": "function",
                    "function": {
                        "name": tool["name"],
                        "description": tool["description"],
                        "parameters": tool["parameters"]
                    }
                } for tool in tools
            ]

        response = await self.groq_client.chat.completions.create(**params)

        # Extract tool calls
        tool_calls = None
        if response.choices[0].message.tool_calls:
            tool_calls = [
                {
                    "id": tc.id,
                    "name": tc.function.name,
                    "arguments": tc.function.arguments
                } for tc in response.choices[0].message.tool_calls
            ]

        return {
            "content": response.choices[0].message.content or "",
            "model": "llama-3.3-70b-versatile",
            "provider": "groq",
            "tool_calls": tool_calls,
            "usage": {
                "prompt_tokens": response.usage.prompt_tokens,
                "completion_tokens": response.usage.completion_tokens,
                "total_tokens": response.usage.total_tokens
            },
            "finish_reason": response.choices[0].finish_reason,
            "cost_usd": 0.0  # Groq is free
        }

    async def _call_gemini(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: Optional[int]
    ) -> Dict[str, Any]:
        """Call Gemini API"""
        # Convert messages to Gemini format
        gemini_messages = []
        for msg in messages:
            if msg["role"] == "system":
                # Gemini doesn't have system role, prepend to first user message
                continue
            gemini_messages.append({
                "role": "user" if msg["role"] == "user" else "model",
                "parts": [msg["content"]]
            })

        # Generate response
        response = await self.gemini_model.generate_content_async(
            gemini_messages,
            generation_config={
                "temperature": temperature,
                "max_output_tokens": max_tokens or 2048,
            }
        )

        # Estimate tokens (Gemini doesn't provide exact counts)
        prompt_tokens = sum(len(msg["content"].split()) for msg in messages) * 1.3
        completion_tokens = len(response.text.split()) * 1.3
        total_tokens = int(prompt_tokens + completion_tokens)

        # Calculate cost (Gemini 1.5 Pro pricing)
        cost_usd = (prompt_tokens * 1.25 / 1_000_000) + (completion_tokens * 5.0 / 1_000_000)

        return {
            "content": response.text,
            "model": "gemini-1.5-pro",
            "provider": "gemini",
            "tool_calls": None,  # Gemini tool calling not implemented yet
            "usage": {
                "prompt_tokens": int(prompt_tokens),
                "completion_tokens": int(completion_tokens),
                "total_tokens": total_tokens
            },
            "finish_reason": "stop",
            "cost_usd": cost_usd
        }

File: backend/src/services/aiProviderService.ts
import OpenAI from 'openai';
import { Groq } from 'groq-sdk';
import { GoogleGenerativeAI } from '@google/generative-ai';
import Anthropic from '@anthropic-ai/sdk';
import { logger } from '../utils/logger';

export interface AIResponse {
  content: string;
  provider: string;
  model: string;
  tokens: number;
  responseTime: number;
  success: boolean;
  error?: string;
}

export interface ChatMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

export class AIProviderService {
  private openai: OpenAI | null = null;
  private groq: Groq | null = null;
  private gemini: GoogleGenerativeAI | null = null;
  private anthropic: Anthropic | null = null;

  constructor() {
    this.initializeProviders();
  }

  private initializeProviders(): void {
    // Initialize OpenAI
    if (process.env.OPENAI_API_KEY) {
      this.openai = new OpenAI({
        apiKey: process.env.OPENAI_API_KEY,
      });
      logger.info('OpenAI provider initialized');
    }

    // Initialize Groq
    if (process.env.GROQ_API_KEY) {
      this.groq = new Groq({
        apiKey: process.env.GROQ_API_KEY,
      });
      logger.info('Groq provider initialized');
    }

    // Initialize Gemini
    if (process.env.GEMINI_API_KEY) {
      this.gemini = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
      logger.info('Gemini provider initialized');
    }

    // Initialize Anthropic
    if (process.env.ANTHROPIC_API_KEY) {
      this.anthropic = new Anthropic({
        apiKey: process.env.ANTHROPIC_API_KEY,
      });
      logger.info('Anthropic provider initialized');
    }
  }

  private async callOpenAI(messages: ChatMessage[]): Promise<AIResponse> {
    if (!this.openai) {
      throw new Error('OpenAI provider not initialized');
    }

    const startTime = Date.now();

    try {
      const completion = await this.openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: messages,
        max_tokens: 1000,
        temperature: 0.7,
      });

      const responseTime = Date.now() - startTime;

      return {
        content: completion.choices[0]?.message?.content || 'No response generated',
        provider: 'openai',
        model: 'gpt-4o-mini',
        tokens: completion.usage?.total_tokens || 0,
        responseTime,
        success: true,
      };
    } catch (error) {
      const responseTime = Date.now() - startTime;
      logger.error('OpenAI API error:', error);
      return {
        content: '',
        provider: 'openai',
        model: 'gpt-4o-mini',
        tokens: 0,
        responseTime,
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  private async callGroq(messages: ChatMessage[]): Promise<AIResponse> {
    if (!this.groq) {
      throw new Error('Groq provider not initialized');
    }

    const startTime = Date.now();

    try {
      const completion = await this.groq.chat.completions.create({
        model: 'llama-3.3-70b-versatile',
        messages: messages,
        max_tokens: 1000,
        temperature: 0.7,
      });

      const responseTime = Date.now() - startTime;

      return {
        content: completion.choices[0]?.message?.content || 'No response generated',
        provider: 'groq',
        model: 'llama-3.3-70b-versatile',
        tokens: completion.usage?.total_tokens || 0,
        responseTime,
        success: true,
      };
    } catch (error) {
      const responseTime = Date.now() - startTime;
      logger.error('Groq API error:', error);
      return {
        content: '',
        provider: 'groq',
        model: 'llama-3.3-70b-versatile',
        tokens: 0,
        responseTime,
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  private async callGemini(messages: ChatMessage[]): Promise<AIResponse> {
    if (!this.gemini) {
      throw new Error('Gemini provider not initialized');
    }

    const startTime = Date.now();

    try {
      const model = this.gemini.getGenerativeModel({ model: 'gemini-1.5-pro' });

      // Convert messages to Gemini format
      const geminiMessages = messages
        .filter(msg => msg.role !== 'system')
        .map(msg => ({
          role: msg.role === 'user' ? 'user' : 'model',
          parts: [{ text: msg.content }],
        }));

      const chat = model.startChat({
        history: geminiMessages.slice(0, -1),
      });

      const prompt = messages[messages.length - 1].content;
      const result = await chat.sendMessage(prompt);
      const response = await result.response;

      const responseTime = Date.now() - startTime;

      return {
        content: response.text(),
        provider: 'gemini',
        model: 'gemini-1.5-pro',
        tokens: response.text().length, // Approximate token count
        responseTime,
        success: true,
      };
    } catch (error) {
      const responseTime = Date.now() - startTime;
      logger.error('Gemini API error:', error);
      return {
        content: '',
        provider: 'gemini',
        model: 'gemini-1.5-pro',
        tokens: 0,
        responseTime,
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  private async callAnthropic(messages: ChatMessage[]): Promise<AIResponse> {
    if (!this.anthropic) {
      throw new Error('Anthropic provider not initialized');
    }

    const startTime = Date.now();

    try {
      // Convert messages to Anthropic format
      const anthropicMessages = messages
        .filter(msg => msg.role !== 'system')
        .map(msg => ({
          role: msg.role === 'user' ? 'user' : 'assistant',
          content: msg.content,
        }));

      const systemMessage = messages.find(msg => msg.role === 'system');

      const completion = await this.anthropic.messages.create({
        model: 'claude-3-haiku-20240307',
        max_tokens: 1000,
        temperature: 0.7,
        system: systemMessage?.content,
        messages: anthropicMessages,
      });

      const responseTime = Date.now() - startTime;

      return {
        content: completion.content[0]?.type === 'text' ? completion.content[0].text : 'No response generated',
        provider: 'anthropic',
        model: 'claude-3-haiku-20240307',
        tokens: completion.usage.input_tokens + completion.usage.output_tokens,
        responseTime,
        success: true,
      };
    } catch (error) {
      const responseTime = Date.now() - startTime;
      logger.error('Anthropic API error:', error);
      return {
        content: '',
        provider: 'anthropic',
        model: 'claude-3-haiku-20240307',
        tokens: 0,
        responseTime,
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  public async generateResponse(
    userMessage: string,
    conversationHistory: ChatMessage[] = [],
    preferredProvider: string = 'openai'
  ): Promise<AIResponse> {
    // Build system prompt for Hebrew/English fitness assistant
    const systemPrompt = `You are SweatBot, a Hebrew/English fitness AI assistant. You help users with:
1. Workout planning and exercise recommendations
2. Fitness goal setting and tracking
3. Exercise form and technique guidance
4. Motivation and support

Guidelines:
- Respond in the user's language (Hebrew or English)
- Be encouraging and supportive
- Provide specific, actionable advice
- Ask clarifying questions when needed
- Include exercise safety considerations
- Recommend consulting professionals for medical conditions

Keep responses concise but informative, under 200 words when possible.`;

    const messages: ChatMessage[] = [
      { role: 'system', content: systemPrompt },
      ...conversationHistory.slice(-10), // Keep last 10 messages for context
      { role: 'user', content: userMessage }
    ];

    // Try preferred provider first, then fallback to others
    const providers = this.getProviderOrder(preferredProvider);

    for (const provider of providers) {
      try {
        let response: AIResponse;

        switch (provider) {
          case 'openai':
            response = await this.callOpenAI(messages);
            break;
          case 'groq':
            response = await this.callGroq(messages);
            break;
          case 'gemini':
            response = await this.callGemini(messages);
            break;
          case 'anthropic':
            response = await this.callAnthropic(messages);
            break;
          default:
            throw new Error(`Unknown provider: ${provider}`);
        }

        if (response.success) {
          logger.info(`AI response generated using ${provider}`, {
            tokens: response.tokens,
            responseTime: response.responseTime,
          });
          return response;
        } else {
          logger.warn(`Provider ${provider} failed:`, response.error);
        }
      } catch (error) {
        logger.warn(`Provider ${provider} error:`, error);
      }
    }

    // All providers failed
    return {
      content: 'I apologize, but I\'m having trouble connecting to my AI services right now. Please try again in a moment.',
      provider: 'none',
      model: '',
      tokens: 0,
      responseTime: 0,
      success: false,
      error: 'All AI providers failed',
    };
  }

  private getProviderOrder(preferred: string): string[] {
    const availableProviders: string[] = [];

    if (this.openai) availableProviders.push('openai');
    if (this.groq) availableProviders.push('groq');
    if (this.gemini) availableProviders.push('gemini');
    if (this.anthropic) availableProviders.push('anthropic');

    // Put preferred provider first if available
    if (availableProviders.includes(preferred)) {
      const ordered = [preferred];
      ordered.push(...availableProviders.filter(p => p !== preferred));
      return ordered;
    }

    return availableProviders;
  }

  public async healthCheck(): Promise<{ provider: string; status: string }> {
    const healthChecks = [];

    if (this.openai) {
      try {
        await this.openai.models.list();
        healthChecks.push({ provider: 'openai', status: 'healthy' });
      } catch (error) {
        healthChecks.push({ provider: 'openai', status: 'unhealthy' });
      }
    }

    if (this.groq) {
      try {
        await this.groq.models.list();
        healthChecks.push({ provider: 'groq', status: 'healthy' });
      } catch (error) {
        healthChecks.push({ provider: 'groq', status: 'unhealthy' });
      }
    }

    if (this.gemini) {
      try {
        const model = this.gemini.getGenerativeModel({ model: 'gemini-1.5-pro' });
        await model.generateContent('Health check');
        healthChecks.push({ provider: 'gemini', status: 'healthy' });
      } catch (error) {
        healthChecks.push({ provider: 'gemini', status: 'unhealthy' });
      }
    }

    if (this.anthropic) {
      try {
        await this.anthropic.models.list();
        healthChecks.push({ provider: 'anthropic', status: 'healthy' });
      } catch (error) {
        healthChecks.push({ provider: 'anthropic', status: 'unhealthy' });
      }
    }

    return healthChecks;
  }
}

// Export singleton instance
export const aiProviderService = new AIProviderService();